### PR TITLE
Brand-new crooshie AI

### PR DESCRIPTION
**Pull request by xDanilcusx**
# Fuck it, we croosh!
### A.K.A shotgun pen nerf, crusher amalgameton and many, many, too fucking many effects

### CODE REVIEW

**Comment by morrowwolf** [comment]
Alright gonna test in a live environment.

Initial thoughts:

Definitely not a fan of the animation. It should be properly sprited or use some other way to portray it. Manipulating the sprite like this looks pretty weird and doesn't work on north-south dir.

File: code/modules/mob/living/carbon/xenomorph/ai/xeno_ai.dm [hunk 1]
```diff
@@ -165,7 +165,7 @@
 
 	if(!current_path || (next_path_generation < world.time && current_target_turf != T))
 		if(!XENO_CALCULATING_PATH(src) || current_target_turf != T)
-			SSxeno_pathfinding.calculate_path(src, T, max_range, src, CALLBACK(src, PROC_REF(set_path)), list(src, current_target))
+			SSxeno_pathfinding.calculate_path(src, T, max_range, src, CALLBACK(src, PROC_REF(set_path)), list(src, current_target) + ignore)
```
Comment (fira) [inline code/modules/mob/living/carbon/xenomorph/ai/xeno_ai.dm:168]: Doing this directly adds a stray `null` in the list if `ignore` is not set, you probably want to either check against that or init it to list()

File: code/modules/mob/living/carbon/xenomorph/castes/Crusher.dm [hunk 2]
```diff
@@ -56,19 +59,39 @@
...
+	behavior_delegate = new /datum/behavior_delegate/crusher_charger()
+	behavior_delegate.bound_xeno = src
+	behavior_delegate.add_to_xeno()
+
+	RegisterSignal(src, COMSIG_MOB_PRE_CLICK, PROC_REF(on_click))
+
+/mob/living/carbon/xenomorph/crusher/proc/on_click(mob/living/carbon/xenomorph/X, atom/target, list/mods)
+	SIGNAL_HANDLER
+	if(HAS_TRAIT(src, TRAIT_CHARGING) && !istype(target, /atom/movable/screen))
+		return COMPONENT_INTERRUPT_CLICK
```
Comment (fira) [inline code/modules/mob/living/carbon/xenomorph/castes/Crusher.dm:93]: What for?

File: code/modules/mob/living/carbon/xenomorph/ai/xeno_ai.dm (hunk 1 above)
Comment (morrowwolf) [reply code/modules/mob/living/carbon/xenomorph/ai/xeno_ai.dm:175]: https://github.com/PvE-CMSS13/PvE-CMSS13/pull/51/commits/901c6543ab6392bbc18f64bf87116b403b46da90

File: code/modules/mob/living/carbon/xenomorph/ai/movement/crusher.dm [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+
+#define AI_NEW_TARGET_COOLDOWN 1 SECONDS
+
+/datum/xeno_ai_movement/crusher/New(mob/living/carbon/xenomorph/parent)
+	. = ..()
+	addtimer(CALLBACK(src, PROC_REF(get_new_target), parent), AI_NEW_TARGET_COOLDOWN, TIMER_UNIQUE|TIMER_LOOP|TIMER_DELETE_ME)
+
+/datum/xeno_ai_movement/crusher/proc/get_new_target(mob/living/carbon/xenomorph/parent)
+	parent.current_target = parent.get_target(parent.ai_range)
```
[the remark this answers is not in the document]
Comment (morrowwolf) [reply code/modules/mob/living/carbon/xenomorph/ai/movement/crusher.dm]: https://github.com/PvE-CMSS13/PvE-CMSS13/pull/51/commits/2e062cbd02866dd4e8539cddcd9742bacafc6e6e

File: code/modules/mob/living/carbon/xenomorph/castes/Crusher.dm [hunk 4]
```diff
@@ -56,19 +59,39 @@
...
 
 	icon_xeno = 'icons/mob/xenos/crusher.dmi'
 	icon_xenonid = 'icons/mob/xenonids/crusher.dmi'
 
+	ai_range = 28
+
+/mob/living/carbon/xenomorph/crusher/init_movement_handler()
+	return new /datum/xeno_ai_movement/crusher(src)
+
+/mob/living/carbon/xenomorph/crusher/New(loc, ...)
```
[the remark this answers is not in the document]
Comment (morrowwolf) [reply code/modules/mob/living/carbon/xenomorph/castes/Crusher.dm]: https://github.com/PvE-CMSS13/PvE-CMSS13/pull/51/commits/5891e7afe58e59645a4eceaea4b77efee048f0a4

File: code/modules/mob/living/carbon/xenomorph/castes/Crusher.dm [hunk 5]
```diff
@@ -56,19 +59,39 @@
...
+	. = ..()
+	qdel(behavior_delegate)
+	behavior_delegate = new /datum/behavior_delegate/crusher_charger()
+	behavior_delegate.bound_xeno = src
+	behavior_delegate.add_to_xeno()
```
[the remark this answers is not in the document]
Comment (morrowwolf) [reply code/modules/mob/living/carbon/xenomorph/castes/Crusher.dm]: https://github.com/PvE-CMSS13/PvE-CMSS13/pull/51/commits/5891e7afe58e59645a4eceaea4b77efee048f0a4

File: code/modules/mob/living/carbon/xenomorph/castes/Crusher.dm (hunk 2 above)
Comment (morrowwolf) [reply code/modules/mob/living/carbon/xenomorph/castes/Crusher.dm]: https://github.com/PvE-CMSS13/PvE-CMSS13/pull/51/commits/5891e7afe58e59645a4eceaea4b77efee048f0a4

File: code/modules/mob/living/carbon/xenomorph/ai/movement/crusher.dm [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+		if(get_dist(target, moving_xeno) >= ai_range)
+			toggle_charging(FALSE)
+			return TRUE
+
+		var/list/to_ram = list()
+		for(var/turf/turfs in getline2(moving_xeno, charge_turf))
+			to_ram += turfs.contents
+
+		var/charging_dir = get_dir(moving_xeno, charge_turf)
+		if(moving_xeno.move_to_next_turf(charge_turf, world.maxx, to_ram))
```
[the remark this answers is not in the document]
Comment (morrowwolf) [reply code/modules/mob/living/carbon/xenomorph/ai/movement/crusher.dm]: https://github.com/PvE-CMSS13/PvE-CMSS13/pull/51/commits/0705e5727608f009f993327714e13a65c7875845

**Comment by fira** [comment]
👍

**Comment by morrowwolf** [comment]
Crushers hitting a head on charge is going to be uuuh terrifying with this. We'll see if people like it or not iunno

**Comment by fira** [comment]
👍

**Review by morrowwolf — APPROVED** [review]
Gonna look over again tomorrow just to make sure I'm not missing something but good probably

**Comment by morrowwolf** [comment]
Oh Fira +1'd it we're good

**Pull request merged**